src/service.rs

```diff
@@ -551,10 +555,31 @@ where
         }
         let (pub_key, sign_key) = self.builder_keys.clone();
 
-        if let Some(block_info) = self.global_state.write_arc().await.blocks.get(&block_id) {
+        let extracted_block_info_option = {
+            // We store this write lock guard separately to make it explicit
+            // that this will end up holding a lock for the duration of this
+            // closure.
+            //
+            // Additionally, we clone the properties from the block_info that
+            // end up being cloned if found anyway.  Since we know this already
+            // we can perform the clone here to avoid holding the lock for
+            // longer than needed.
+            let mut global_state_write_lock_guard = self.global_state.write_arc().await;
+            let block_info_some = global_state_write_lock_guard.blocks.get(&block_id);
+
+            block_info_some.map(|block_info| {
+                (
+                    block_info.vid_trigger.clone(),
+                    block_info.block_payload.clone(),
+                    block_info.metadata.clone(),
+                )
+            })
+        };
+
+        if let Some((vid_trigger, block_payload, metadata)) = extracted_block_info_option {
             tracing::info!("Trying sending vid trigger info for {block_id}",);
 
-            if let Some(trigger_writer) = block_info.vid_trigger.write().await.take() {
+            if let Some(trigger_writer) = vid_trigger.write().await.take() {
```

Comment:
Since `vid_trigger` is obtained by a `clone`, I'm wondering if this would cause the `take` operation not applied to the original `block_info` and `global_state`.

Reply:
I believe this *should* be safe as we are actually cloning the `Arc`, which should hold a reference to the same instance of the contained value.  This should mean that the `take` is actually sitll on the original value itself.